cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,308 @@
+package db
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/ledgerbucketwindow"
+)
+
+const transactionTableName = "transactions"
+
+var ErrNoTransaction = errors.New("no transaction with this hash exists")
+
+type Transaction struct {
+	Result           []byte   // XDR encoded xdr.TransactionResult
+	Meta             []byte   // XDR encoded xdr.TransactionMeta
+	Envelope         []byte   // XDR encoded xdr.TransactionEnvelope
+	Events           [][]byte // XDR encoded xdr.DiagnosticEvent
+	FeeBump          bool
+	ApplicationOrder int32
+	Successful       bool
+	Ledger           ledgerbucketwindow.LedgerInfo
+}
+
+// TransactionWriter is used during ingestion to write LCM.
+type TransactionWriter interface {
+	InsertTransactions(lcm xdr.LedgerCloseMeta) error
+	RegisterMetrics(ingest, insert, count prometheus.Observer)
+}
+
+// TransactionReader provides all of the public ways to read from the DB.
+type TransactionReader interface {
+	GetTransaction(ctx context.Context, hash xdr.Hash) (Transaction, ledgerbucketwindow.LedgerRange, error)
+	GetLedgerRange(ctx context.Context) (ledgerbucketwindow.LedgerRange, error)
+}
+
+type transactionHandler struct {
+	log        *log.Entry
+	db         db.SessionInterface
+	stmtCache  *sq.StmtCache
+	passphrase string
+
+	ingestMetric, insertMetric, countMetric prometheus.Observer
+}
+
+func NewTransactionReader(log *log.Entry, db db.SessionInterface, passphrase string) TransactionReader {
+	return &transactionHandler{log: log, db: db, passphrase: passphrase}
+}
+
+func (txn *transactionHandler) InsertTransactions(lcm xdr.LedgerCloseMeta) error {
+	start, mid := time.Now(), time.Now()
+	txCount := lcm.CountTransactions()
+	L := txn.log.
+		WithField("ledger_seq", lcm.LedgerSequence()).
+		WithField("tx_count", txCount)
+
+	defer func() {
+		if txn.ingestMetric != nil {
+			txn.ingestMetric.Observe(time.Since(start).Seconds())
+			txn.insertMetric.Observe(time.Since(mid).Seconds())
```

Comment:
it seems like these two metrics are measuring the same values?

Reply:
ok, I see that `mid` gets updated later in the function. In practice I think the time spent collecting the transactions from the ledger close meta will be extremely small so having two different metrics is probably overkill

Reply:
`mid` was specifically to measure the sql/db portion of the ingestion, but I do agree with you that it's probably going to be nearly identical to `start` to which the distinction probably isn't worth it :+1: done in 9fc64a9